frontend/src/main/scala/bloop/bsp/BloopBspServices.scala

```diff
@@ -525,37 +526,65 @@ final class BloopBspServices(
   }
 
   def run(params: bsp.RunParams): BspEndpointResponse[bsp.RunResult] = {
+    def getMainClass(project: Project, state: State): Either[Exception, bsp.ScalaMainClass] = {
+      params.dataKind match {
+        // TODO replace with RunParamsDataKind.ScalaMainClass once bsp updates
+        case Some("scala-main-class") =>
+          params.data match {
+            case None =>
+              Left(new IllegalStateException(s"Missing data for $params"))
+            case Some(json) =>
+              json.as[bsp.ScalaMainClass]
+          }
+        case Some(kind) =>
+          Left(new IllegalArgumentException(s"Unsupported data kind: $kind"))
+        case None =>
+          val cmd = Commands.Run(List(project.name))
+          Interpreter.getMainClass(state, project, cmd.main) match {
+            case Right(name) =>
+              Right(new bsp.ScalaMainClass(name, cmd.args, Nil))
+            case Left(_) =>
+              Left(new IllegalStateException(s"Main class for project $project not found"))
+          }
+      }
+    }
+
     def run(
         id: BuildTargetIdentifier,
         project: Project,
         state: State
     ): Task[State] = {
       import bloop.engine.tasks.LinkTask.{linkMainWithJs, linkMainWithNative}
       val cwd = state.commonOptions.workingPath
-      val cmd = Commands.Run(List(project.name))
-      Interpreter.getMainClass(state, project, cmd.main) match {
-        case Left(state) =>
-          Task.now(sys.error(s"Failed to run main class in ${project.name}"))
+
+      getMainClass(project, state) match {
+        case Left(error) =>
+          Task.now(sys.error(s"Failed to run main class in $project due to: ${error.getMessage}"))
```

Comment:
Will it show up for the client as in `logger.showWarning`?

Reply:
It should be shown to the client in the logs, but I'm not sure this will be shown via bap's showMessage. Should this be Metals responsibility?